lib/rails_multisite/connection_management.rb

```diff
@@ -104,24 +108,19 @@ def self.current_db
     end
 
     def self.current_hostname
-      spec = @instance.connection_spec(db: self.current_db) if @instance
-      spec ||= ActiveRecord::Base.connection_pool.spec
-      config = spec.config
-      config[:host_names].nil? ? config[:host] : config[:host_names].first
+      current_db_hostnames.first
     end
 
     def self.current_db_hostnames
-      spec = @instance.connection_spec(db: self.current_db) if @instance
-      spec ||= ActiveRecord::Base.connection_pool.spec
-      config = spec.config
+      config = (@instance&.connection_spec(db: current_db) || ConnectionSpecification.current).config
       config[:host_names].nil? ? [config[:host]] : config[:host_names]
     end
 
     def self.connection_spec(opts)
       if @instance
```

Comment:
Tiny suggestion but this could be a one-liner? (unless there's a performance difference between using `if / else` and `&. ||`)

```suggestion
      @instance&.connection_spec(opts) || ConnectionSpecification.current
```

Reply:
Yes you’re totally right, I’m not 100% sure but I guess the performance penalty (if there’s any) should be small enough to make the change :+1: 
We could also apply this logic to a lot of areas of this class I think but I didn’t change too much things at once :sweat_smile:

Reply:
Ah actually it breaks specs :grimacing:

Reply:
Ah yes I remember now :man_facepalming: There’s a slight subtlety with this: when the instance exists but there’s no connection spec for the given opts the current code will return `nil`. With the change it will actually return `ConnectionSpecification.current` which will break the expected behavior.

Reply:
But yes we see this pattern (`if @instance else …`) a lot, a way to refactor this could be by using some sort of wrapper maybe, one wrapping `@instance` and another when `@instance` is not set. Then we just delegate most of the methods to the wrapper that has been chosen. Just thinking out loud here though :grin:

Reply:
Quickly made this: https://github.com/Flink/rails_multisite/pull/1 (still a draft, it’s not completely finished yet) :slightly_smiling_face:

Reply:
Proper PR: https://github.com/discourse/rails_multisite/pull/21 :slightly_smiling_face: